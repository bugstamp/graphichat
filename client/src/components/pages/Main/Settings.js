import React, {
  PureComponent,
  createRef,
} from 'react';
import PropTypes from 'prop-types';
import { map, omit } from 'lodash';

import TextField from '@material-ui/core/TextField';

import {
  SettingWrapper,
  AvatarWrapper,
  Avatar,
  Camera,
  FormWrapper,
} from './styled';

import { getAvatar } from '../../../helpers';
import { meProps } from '../../propTypes';

class Settings extends PureComponent {
  avatarInput = createRef()

  state = {
    errors: {},
    values: {},
    called: {},
  }

  fields = [
    {
      id: 'username',
      label: 'User Name',
    },
    {
      id: 'firstName',
      label: 'First Name',
    },
    {
      id: 'lastName',
      label: 'Last Name',
    },
  ]

  getSnapshotBeforeUpdate(prevProps) {
    const { error } = this.props;
    const snapshot = {
      error: null,
    };

    if (!prevProps.error && error) {
      const { graphQLErrors } = error;
      const { data = null } = graphQLErrors[0];

      if (data) {
        const { message } = graphQLErrors[0];
        const { invalidField = null } = data;

        if (invalidField) {
          snapshot.error = { [invalidField]: message };
        }
      }
    }
    return snapshot;
  }

  componentDidUpdate(prevProps, prevState, { error }) {
    if (error) {
      this.addError(error);
    }
  }

  onChange = ({ target }) => {
    const { name, value } = target;

    this.setState(({ called, values }) => {
      const updatedCalled = !called[name]
        ? { ...called, [name]: true }
        : called;
      const updatedValues = { ...values, [name]: value };

      return {
        called: updatedCalled,
        values: updatedValues,
      };
    });
  }

  onChangeAvatar = (e) => {
    const {
      target: {
        files: [file],
        validity: { valid },
      },
    } = e;
    const { uploadAvatar } = this.props;

    if (valid) {
      uploadAvatar({ variables: { file } });
    }
  }

  onAvatarClick = () => {
    if (this.avatarInput) {
      this.avatarInput.current.click();
    }
  }

  onSubmit = ({ target }) => {
    const { errors } = this.state;
    const { updateUser, me } = this.props;
    const { name: field, value } = target;
    const existedValue = me[field];

    if (value !== existedValue && value) {
      updateUser({ variables: { field, value } });

      if (errors[field]) {
        this.setState(() => ({
          errors: omit(errors, field),
        }));
      }
    } else if (!value) {
      this.onChange({ target: { name: field, value: existedValue } });
    }
  }

  onKeyPressed = (e) => {
    const { key, target } = e;

    if (key === 'Enter') {
      target.blur();
    }
  }

  addError = (error) => {
    this.setState(({ errors }) => ({
      errors: { ...errors, ...error },
    }));
  }

  render() {
    const { called, values, errors } = this.state;
    const { me } = this.props;
    const avatar = getAvatar(me, 'md');

    return (
      <SettingWrapper>
        <AvatarWrapper>
          <input
            ref={this.avatarInput}
            type="file"
            accept="image/png, image/jpg, image/jpeg"
            onChange={this.onChangeAvatar}
            style={{ display: 'none' }}
          />
          <Avatar
            src={avatar.src}
            alt={avatar.text}
            onClick={this.onAvatarClick}
          >
            <Camera fontSize="large" />
          </Avatar>
        </AvatarWrapper>
        <FormWrapper>
          {
            map(this.fields, ({ id, label }) => {
              const value = called[id] ? values[id] : me[id];
              const error = !!errors[id];
              const message = errors[id];

              return (
                <TextField
                  key={id}
                  name={id}
                  value={value}
                  label={label}
                  error={error}
                  helperText={message}
                  variant="outlined"
                  margin="dense"
                  onChange={this.onChange}
                  onBlur={this.onSubmit}
                  onKeyDown={this.onKeyPressed}
                  fullWidth
                  required
                />
              );
            })
          }
        </FormWrapper>
      </SettingWrapper>
    );
  }
}

Settings.defaultProps = {
  me: {},
  error: null,
};
Settings.propTypes = {
  me: PropTypes.shape(meProps),
  uploadAvatar: PropTypes.func.isRequired,
  updateUser: PropTypes.func.isRequired,
  error: PropTypes.instanceOf(Error),
};

export default Settings;
