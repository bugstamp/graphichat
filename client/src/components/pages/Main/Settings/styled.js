import styled, { keyframes } from 'styled-components';
import { position } from 'polished';
import { fadeOut } from 'react-animations';

import MaterialAvatar from '@material-ui/core/Avatar';
import CameraIcon from '@material-ui/icons/CameraAlt';
import InputAdornment from '@material-ui/core/InputAdornment';

import { getStyledProps, getSpacing } from '../../../../styles';

const fadeOutAnimation = keyframes`${fadeOut}`;

const SettingsWrapper = styled.div`
  width: 100%;
  display: flex;
  justify-content: center;

  ${(props) => {
    const breakpoints = getStyledProps('theme.breakpoints')(props);
    const smDown = breakpoints.down('xs');

    return `
      ${smDown} {
        flex-flow: column;
        justify-content: flex-start;
      }
    `;
  }}
`;

const InputCheckIconWrapper = styled(InputAdornment)`
  && {
    animation: ${fadeOutAnimation};
    animation-duration: .5s;
    animation-delay: 5s;
    animation-fill-mode: forwards;
  }
`;

const AvatarWrapper = styled.div`
  position: relative;
  display: flex;
`;

const Avatar = styled(MaterialAvatar)`
  && {
    width: 172px;
    height: 172px;
    margin: 0 auto;
    position: relative;
    color: ${({ src }) => (src && '#fff')};
    cursor: pointer;

    ${({ src }) => src && `
      > svg {
        opacity: 0;
      }

      &:hover {
        > svg {
          opacity: .6;
        }
      }
    `}
  }
`;

const Camera = styled(CameraIcon)`
  && {
    ${position('absolute', '50%', null, null, '50%')};
    transform: translate(-50%, -50%);
    transition: all .3s ease;
  }
`;

const FormWrapper = styled.div`
  width: 100%;
  padding: ${getSpacing(1)};
`;

export {
  SettingsWrapper,
  AvatarWrapper,
  Avatar,
  Camera,
  FormWrapper,
  InputCheckIconWrapper,
};
