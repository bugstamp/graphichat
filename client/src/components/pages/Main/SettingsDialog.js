import React, { memo } from 'react';
import PropTypes from 'prop-types';

import ResponsiveDialog from '../../common/ResponsiveDialog/ResponsiveDialog';
import Settings from './Settings';

import SettingsDialogContainer from '../../containers/SettingsDialogContainer';

const SettingsDialog = ({ open, toggle }) => {
  return (
    <SettingsDialogContainer>
      {({
        getMe: {
          data = {},
        },
        uploadAvatar: {
          mutation: uploadAvatar,
          result: {
            loading: avatarUploading,
          },
        },
        updateUser: {
          mutation: updateUser,
          result: {
            loading: updating,
            error,
          },
        },
      }) => {
        const { me } = data;
        const loading = avatarUploading || updating;

        return (
          <ResponsiveDialog
            open={open}
            toggle={toggle}
            loading={loading}
            title="Settings"
          >
            <Settings
              me={me}
              error={error}
              updateUser={updateUser}
              uploadAvatar={uploadAvatar}
            />
          </ResponsiveDialog>
        );
      }}
    </SettingsDialogContainer>
  );
};

SettingsDialog.defaultProps = {
  open: false,
};
SettingsDialog.propTypes = {
  open: PropTypes.bool,
  toggle: PropTypes.func.isRequired,
};

export default memo(SettingsDialog);
