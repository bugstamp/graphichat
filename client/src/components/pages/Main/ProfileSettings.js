import React, { PureComponent, createRef } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { map, set, omit } from 'lodash';
import { position } from 'polished';

import MaterialAvatar from '@material-ui/core/Avatar';
import CameraIcon from '@material-ui/icons/CameraAlt';
import TextField from '@material-ui/core/TextField';

import { getSpacing } from '../../../styles';
import { meProps } from '../../propTypes';

const ProfileSettingsWrapper = styled.div`
  display: flex;
`;

const AvatarWrapper = styled.div`
  padding-bottom: ${getSpacing(2)};
  position: relative;
`;

const Avatar = styled(MaterialAvatar)`
  && {
    width: 150px;
    height: 150px;
    margin: 0 auto;
    position: relative;
    color: ${({ src }) => (src && '#fff')};
    cursor: pointer;

    ${({ src }) => src && `
      > svg {
        opacity: 0;
      }

      &:hover {
        > svg {
          opacity: .6;
        }
      }
    `}
  }
`;

const Camera = styled(CameraIcon)`
  && {
    ${position('absolute', '50%', null, null, '50%')};
    transform: translate(-50%, -50%);
    transition: all .3s ease;
  }
`;

const ProfileWrapper = styled.div`
  width: 100%;
  padding: ${getSpacing(1)};
`;

class ProfileSettings extends PureComponent {
  avatarInput = createRef()

  state = {
    errors: {},
    values: {},
    called: {},
  }

  fields = [
    {
      id: 'username',
      label: 'Username',
    },
    {
      id: 'firstName',
      label: 'First Name',
    },
    {
      id: 'lastName',
      label: 'Last Name',
    },
  ]

  getSnapshotBeforeUpdate(prevProps) {
    const { error } = this.props;
    const snapshot = {
      error: null,
    };

    if (!prevProps.error && error) {
      const { graphQLErrors } = error;
      const { message, data = null } = graphQLErrors[0];

      if (data) {
        const { invalidField = null } = data;

        if (invalidField) {
          snapshot.error = { [invalidField]: message };
        }
      }
    }
    return snapshot;
  }

  componentDidUpdate(prevProps, prevState, { error }) {
    if (error) {
      this.addError(error);
    }
  }

  onChange = ({ target }) => {
    const { name, value } = target;

    this.setState(({ called, values }) => {
      if (!called[name]) {
        set(called, name, true);
      }
      return ({
        called,
        values: { ...values, [name]: value },
      });
    });
  }

  onChangeAvatar = ({
    target: {
      files: [file],
      validity: { valid },
    },
  }) => {
    const { uploadAvatar } = this.props;

    if (valid) {
      uploadAvatar({ variables: { file } });
    }
  }

  onAvatarClick = () => {
    if (this.avatarInput) {
      this.avatarInput.current.click();
    }
  }

  onSubmit = ({ target }) => {
    const { errors } = this.state;
    const { updateUser, me } = this.props;
    const { name: field, value } = target;
    const form = { field, value };
    const existedValue = me[field];

    if (value !== existedValue && value) {
      updateUser({ variables: { form } });

      if (errors[field]) {
        this.setState(() => ({
          errors: omit(errors, field),
        }));
      }
    } else if (!value) {
      this.onChange({ target: { name: field, value: existedValue } });
    }
  }

  onKeyPressed = (e) => {
    const { key, target } = e;

    if (key === 'Enter') {
      target.blur();
    }
  }

  addError = (error) => {
    this.setState(({ errors }) => ({
      errors: { ...errors, ...error },
    }));
  }

  render() {
    const { called, values, errors } = this.state;
    const { avatar, me } = this.props;

    return (
      <ProfileSettingsWrapper>
        <AvatarWrapper>
          <input
            ref={this.avatarInput}
            type="file"
            accept="image/png, image/jpg, image/jpeg"
            onChange={this.onChangeAvatar}
            style={{ display: 'none' }}
          />
          <Avatar
            src={avatar.src}
            alt={avatar.text}
            onClick={this.onAvatarClick}
          >
            <Camera fontSize="large" />
          </Avatar>
        </AvatarWrapper>
        <ProfileWrapper>
          {
            map(this.fields, ({ id, label }) => {
              const value = called[id] ? values[id] : me[id];
              const error = !!errors[id];
              const message = errors[id];

              return (
                <TextField
                  key={id}
                  name={id}
                  value={value}
                  label={label}
                  error={error}
                  helperText={message}
                  variant="outlined"
                  margin="dense"
                  onChange={this.onChange}
                  onBlur={this.onSubmit}
                  onKeyDown={this.onKeyPressed}
                  fullWidth
                  required
                />
              );
            })
          }
        </ProfileWrapper>
      </ProfileSettingsWrapper>
    );
  }
}

ProfileSettings.defaultProps = {
  me: {},
  error: null,
};
ProfileSettings.propTypes = {
  avatar: PropTypes.shape({
    src: PropTypes.string,
    text: PropTypes.string,
  }).isRequired,
  me: PropTypes.shape(meProps),
  uploadAvatar: PropTypes.func.isRequired,
  updateUser: PropTypes.func.isRequired,
  error: PropTypes.instanceOf(Error),
};

export default ProfileSettings;
