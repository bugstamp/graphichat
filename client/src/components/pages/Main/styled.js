import styled from 'styled-components';

import Grid from '@material-ui/core/Grid';
import Paper from '@material-ui/core/Paper';

import { getStyledProps } from '../../../styles';

const AppGrid = styled(Grid)`
  background-color: ${getStyledProps('theme.palette.grey.300')};
`;

const AppContainer = styled(Paper)`
  && {
    flex: 1 auto;
    display: flex;
    overflow: hidden;

  ${(props) => {
    const breakpoints = getStyledProps('theme.breakpoints')(props);
    const lgUp = breakpoints.up('lg');
    const mdDown = breakpoints.down('md');

    return `
      ${lgUp} {
        margin: 20px 40px;
      }
      ${mdDown} {
        border-radius: 0;
      }
    `;
  }};
  }
`;

export {
  AppGrid,
  AppContainer,
};
