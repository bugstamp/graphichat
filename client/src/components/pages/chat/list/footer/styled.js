import styled from 'styled-components';

import { getSpacing } from '../../../../../styles';

export const ChatListFooterStyled = styled.div`
  display: flex;
  justify-content: center;
  padding: ${getSpacing(2)} ${getSpacing(1)};

  && {
    button {
      width: 50%;
    }
  }
`;

export default null;
