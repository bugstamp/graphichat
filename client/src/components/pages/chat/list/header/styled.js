import styled from 'styled-components';
import { position } from 'polished';

import IconButton from '@material-ui/core/IconButton';

import { getSpacing } from '../../../../../styles';

export const ChatListHeaderStyled = styled.div`
  display: flex;
  flex-flow: column;
  padding: ${getSpacing(2)} ${getSpacing(1)};
`;

export const HeaderTitle = styled.div`
  position: relative;
  margin-bottom: ${getSpacing(1)};
`;

export const AddButton = styled(IconButton)`
  && {
    ${position('absolute', 0, 0, null, null)};
    padding: ${getSpacing(0.5)};
  }
`;
