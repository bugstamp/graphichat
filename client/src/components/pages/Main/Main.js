import React, {
  useState,
  useCallback,
  useEffect,
} from 'react';
import PropTypes from 'prop-types';

import Grid from '@material-ui/core/Grid';
import Hidden from '@material-ui/core/Hidden';

import Navigation from '../../common/Navigation/Navigation';
import SettingsDialog from './SettingsDialog';
import { AppGrid, AppContainer } from './styled';

const Main = (props) => {
  const {
    children,
    loading,
    userId,
    signOut,
    sessionExpired,
  } = props;
  const [settingsDialog, toggleSettingsDialog] = useState(false);

  const handleToggleSettingsDialog = useCallback(() => {
    toggleSettingsDialog(!settingsDialog);
  }, [settingsDialog]);

  const handleSignOut = useCallback(() => {
    signOut({ variables: { userId } });
  }, [userId, signOut]);

  useEffect(() => {
    if (sessionExpired) {
      handleSignOut();
    }
  }, [sessionExpired, handleSignOut]);

  return (
    <AppGrid container spacing={0} justify="center">
      <AppContainer elevation={5}>
        <Grid container spacing={0}>
          <Hidden smDown>
            <Grid item>
              <Navigation
                toggleSettingsDialog={handleToggleSettingsDialog}
                signOut={handleSignOut}
              />
            </Grid>
          </Hidden>
          <Grid item xs>
            {children}
          </Grid>
        </Grid>
      </AppContainer>
      <Hidden mdUp implementation="css">
        <Navigation
          variant="horizontal"
          toggleSettingsDialog={toggleSettingsDialog}
          signOut={signOut}
        />
      </Hidden>
      <SettingsDialog
        open={settingsDialog}
        toggle={handleToggleSettingsDialog}
      />
    </AppGrid>
  );
};

Main.defaultProps = {
  children: null,
  userId: null,
};
Main.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.node]),
  loading: PropTypes.bool.isRequired,
  userId: PropTypes.string,
  sessionExpired: PropTypes.bool.isRequired,
  signOut: PropTypes.func.isRequired,
};

export default Main;
