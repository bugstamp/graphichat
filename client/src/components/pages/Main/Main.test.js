import React from 'react';
import { act } from 'react-dom/test-utils';
import { shallow } from 'enzyme';
import wait from 'waait';

import Main from './Main';
import SettingsDialog from './SettingsDialog';
import { AppGrid } from './styled';

import { me } from '../../../__mocks__/mockedQueryData';
import { mountMockedProvider } from '../../../__mocks__/mockedProvider';

// eslint-disable-next-line
const TestChild = ({ toggleSettingsDialog, signOut }) => (
  <div>
    <button type="button" className="btn1" onClick={toggleSettingsDialog} />
    <button type="button" className="btn2" onClick={signOut} />
  </div>
);

describe('test Main', () => {
  const mockSignOut = jest.fn();
  const defaultProps = {
    loading: false,
    userId: me.id,
    sessionExpired: false,
    signOut: mockSignOut,
  };
  test('snapshot render', () => {
    const wrapper = shallow(<Main {...defaultProps} />);

    expect(wrapper).toMatchSnapshot();
  });
  test('mount render | check default type of children prop', async () => {
    const wrapper = mountMockedProvider((<Main {...defaultProps} />));

    expect(wrapper.find(Main)).toBeTruthy();
    expect(wrapper.find(AppGrid)).toBeTruthy();
    expect(wrapper.find(Main).prop('children')).toBe(null);

    await act(async () => {
      wrapper.update();
    });
  });
  test('pass child | pass props to child clone', async () => {
    const wrapper = mountMockedProvider(<Main {...defaultProps}><TestChild /></Main>);

    expect(wrapper.find(TestChild)).toBeTruthy();
    expect(wrapper.find(TestChild).prop('initialLoading')).toBe(false);
    expect(wrapper.find(TestChild).prop('toggleSettingsDialog')).toBeInstanceOf(Function);
    expect(wrapper.find(TestChild).prop('signOut')).toBeInstanceOf(Function);

    await act(async () => {
      wrapper.update();
    });
  });
  test('toggle settingsDialog', async () => {
    const wrapper = mountMockedProvider(<Main {...defaultProps}><TestChild /></Main>);
    expect(wrapper.find(SettingsDialog).prop('open')).toBe(false);

    await act(async () => {
      await wait();
      wrapper.find(TestChild).find('.btn1').simulate('click');
      wrapper.update();
      expect(wrapper.find(SettingsDialog).prop('open')).toBe(true);
    });
  });
  test('signOut call', async () => {
    const wrapper = mountMockedProvider((
      <Main {...defaultProps}>
        <TestChild />
      </Main>
    ));
    expect(wrapper.find(Main).prop('userId')).toBe(me.id);

    await act(async () => {
      wrapper.find(TestChild).find('.btn2').simulate('click');
      expect(mockSignOut).toBeCalledWith({ variables: { userId: me.id } });
    });
  });
  test('signOut call after session has expired', async () => {
    let wrapper = mountMockedProvider((
      <Main {...defaultProps}>
        <TestChild />
      </Main>
    ));
    expect(wrapper.find(Main).prop('sessionExpired')).toBe(false);

    await act(async () => {
      await wait();
      wrapper = mountMockedProvider((
        <Main {...defaultProps} sessionExpired>
          <TestChild />
        </Main>
      ));
      expect(wrapper.find(Main).prop('sessionExpired')).toBe(true);
      expect(mockSignOut).toBeCalledWith({ variables: { userId: me.id } });
    });
  });
});
