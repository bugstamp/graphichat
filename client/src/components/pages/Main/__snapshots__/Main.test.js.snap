// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Main snapshot render 1`] = `
<styled__AppGrid
  container={true}
  justify="center"
  spacing={0}
>
  <styled__AppContainer>
    <WithStyles(ForwardRef(Grid))
      container={true}
      spacing={0}
    >
      <Hidden
        smDown={true}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
        >
          <Navigation
            itemSize={60}
            logoSize={35}
            signOut={[Function]}
            toggleSettingsDialog={[Function]}
            variant="vertical"
          />
        </WithStyles(ForwardRef(Grid))>
      </Hidden>
      <WithStyles(ForwardRef(Grid))
        item={true}
        xs={true}
      />
    </WithStyles(ForwardRef(Grid))>
  </styled__AppContainer>
  <Memo(SettingsDialog)
    open={false}
    toggle={[Function]}
  />
</styled__AppGrid>
`;
